File: src/leetcode/third/Code_039_CombinationSum.java
package leetcode.third;

import java.util.ArrayList;
import java.util.List;

/**
 * 组合总数
 * 无重复元素 &　元素可以无限次使用
 * 
 * @since 2020-7-18 Saturday 22:02 - 22:15  
 */
public class Code_039_CombinationSum {
    List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(res, new ArrayList<>(), candidates, target, 0, 0);
        return res;
    }

    private void dfs(List<List<Integer>> res, List<Integer> tmp, int[] candidates, int target, int sum, int index) {
        if (sum == target) {
            res.add(new ArrayList<>(tmp));
            return;
        }
        if (sum > target) {
            return;
        }
        for (int i = index; i < candidates.length; i++) {
            tmp.add(candidates[i]);
            dfs(res, tmp, candidates, target, sum + candidates[i], i);
            tmp.remove(tmp.size() - 1);
        }
    }

    public static void main(String[] args) {
        Code_039_CombinationSum c = new Code_039_CombinationSum();
        System.out.println(c.combinationSum(new int[] { 2, 3, 6, 7 }, 7));
        System.out.println(c.combinationSum(new int[] { 2, 3, 5 }, 8));
    }
}

File: src/cpp-leetcode/first/Code_039_CombinationSum.cpp
#include <iostream>
#include <vector>

using std::vector;

/**
 * combination sum
 * 无重复元素的数组 candidates
 * candidates 中的数字可以无限制重复被选取
 *
 * @since 2020-10-28 Wednesday 17:37 - 17:50
 */
class Solution {
    void traceback(vector<vector<int>>& res, vector<int>& tmp, vector<int> candidates, int target, int index) {
        if (target == 0) {
            res.push_back(tmp);
            return;
        }
        for (int i = index; i < candidates.size(); i++) {
            if (target >= candidates[i]) {
                tmp.push_back(candidates[i]);
                traceback(res, tmp, candidates, target - candidates[i], i);
                tmp.pop_back();
            }
        }
    }

   public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> tmp;
        traceback(res, tmp, candidates, target, 0);
        return res;
    }
};

int main(int argc, char const* argv[]) {
    Solution s;
    vector<int> v1 = {2, 3, 5};
    vector<vector<int>> res1 = s.combinationSum(v1, 8);
    for (auto v : res1) {
        std::cout << "[";
        for (int i : v) {
            std::cout << i << " ";
        }
        std::cout << "]\n";
    }
    std::cout << "===\n";
    vector<int> v2 = {2, 3, 6, 7};
    vector<vector<int>> res2 = s.combinationSum(v2, 7);
    for (auto v : res2) {
        std::cout << "[";
        for (int i : v) {
            std::cout << i << " ";
        }
        std::cout << "]\n";
    }
}
